docs/nebula-design-concepts.md

```diff
@@ -0,0 +1,216 @@
+  participate in Raft protocol when it catches up leader.
+
+- Load-balance
+
+      For machines with high stress, migrate the partition they serve to machines
+  with  low stress for better load balancing.
+
+![image][example6]
+
+### Metadata service
+
+The process corresponding to the Meta service is nebula-metad, and its main functions are:
+
+-  User management
+
+      Nebula Graph's user system includes God user, Admin, User, and Guest, each is different in operating privileges.
+
+- Cluster configuration management
+
+      Support online and offline new servers.
+
+- Graph space management
+
+      Add, drop graph space, modify graph space configuration (Raft copies)
+
+-   Schema management
+
+      Nebula Graph is a strong schema database.
+
+      - Types of Tag and Edge properties are recorded by Meta service. Supported types are: int, double, timestamp, list, etc.
+
+      - Multi-version management, supporting adding, modifying and deleting schema, and recording its version.
+
+      - TTL (time-to-live) management, supporting automatic data deletion and space reclamation.
+
+The Meta Service layer is state, and just like the Storage layer, it maintains state persistence through KV Store.
+
+![image][example7]
+
+### Query Engine & Query Language (nGQL)
+
+The process corresponding to the computation layer is nebula-graphd, which consists
+of computing vertices that are fully equal, stateless, and unrelated. There is no
+communication among the computing vertices.
+
+The main function of the Query Engine layer is to analysis the nGQL text sent by the client and generate an execution plan through lexical analysis (Lexer) and parsing (Parser). Then the execution plan will pass to the the execution engine. The execution engine obtains the schema of the graph vertices and edges through the Meta Service, then obtains the data of the vertices and edges through the storage engine layer.
+
+The main optimizations of the Query Engine layer are:
+
+- Asynchronous and concurrent execution
+
+      Since both IO and network are long-latency operations, asynchronous and concurrent operations are required. In addition, to avoid a single long query affecting subsequent queries, a separate resource pool is set for each query to guarantee quality of service (QoS).
+
+- Push down
+
+      Too much data passing back form storage layer to computation layer occupies  too much bandwidth, to avoid that, operators such as conditional filtering (where) are sent to the storage layer along with the query filters.
+
+- Execution plan optimization
+
+      Execution plan optimization in relational database SQL has experienced a long period but research on the optimization of graph query language is relatively
+      less. Nebula Graph also makes some exploration on the execution plan optimization of graph query, including executing plan cache and the concurrent
+      execution of the context-free queries.
+
```

Comment:
Query Optimizer. 
Query optimizer for SQL has experienced a long-term development, but not for graph database. Nebula Graph makes an effort in the optimizations of graph query, including cache executing plan and execute context-free queries in parallel.